lib/Transforms/IPO/StructFieldCacheAnalysis.cpp

```diff
@@ -36,30 +46,672 @@ class StructFieldCacheAnalysisPass : public ModulePass {
+ private:
+  const StructFieldAccessManager* StructManager;
+};
+}
+
+// Functions for StructFieldAccessInfo
+void StructFieldAccessInfo::addFieldAccessNum(const Instruction* I, FieldNumType FieldNum)
+{
+  assert(I->getOpcode() == Instruction::Load || I->getOpcode() == Instruction::Store); // Only loads and stores
+  assert (LoadStoreFieldAccessMap.find(I) == LoadStoreFieldAccessMap.end());
+  LoadStoreFieldAccessMap[I] = FieldNum;
+}
+
+Optional<FieldNumType> StructFieldAccessInfo::getAccessFieldNum(const Instruction* I) const
+{
+  Optional<FieldNumType> ret;
+  auto it = LoadStoreFieldAccessMap.find(I);
+  if (it != LoadStoreFieldAccessMap.end()){
+    ret = it->second;
+  }
+  return ret;
+}
+
+FieldNumType StructFieldAccessInfo::calculateFieldNumFromGEP(const User* U) const
+{
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Calculating field number from GEP: " << *U << "\n");
+  //Operand 0 should be a pointer to the struct
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U));
+  auto* Op = U->getOperand(0);
+  // Make sure Operand 0 is a struct type and matches the current struct type of StructFieldAccessInfo
+  assert(Op->getType()->isPointerTy() && Op->getType()->getPointerElementType()->isStructTy() && Op->getType()->getPointerElementType() == StructureType);
+  if (U->getNumOperands() < 3) // GEP to calculate struct field needs at least 2 indices (operand 1 and 2)
+    return 0;
+  //Operand 1 should be first index to the struct, usually 0; if not 0, it's like goto an element of an array of structs
+  Op = U->getOperand(1);
+  //TODO: ignore this index for now because it's the same for an array of structs
+  assert(Op->getType()->isIntegerTy());
+  //Operand 2 should be the index to the field, and be a constant
+  Op = U->getOperand(2);
+  assert(isa<Constant>(Op));
+  auto* Index = dyn_cast<Constant>(Op);
+  auto Offset = (FieldNumType)Index->getUniqueInteger().getZExtValue();
+  assert(Offset < NumElements);
+  //TODO: ignore indices after this one. If there's indices, the field has to be an array or struct
+  return Offset+1; // return field number starting from 1
+}
+
+void StructFieldAccessInfo::addFieldAccessFromGEP(const User* U)
+{
+  DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Analyze all users of GEP: " << *U << "\n");
+  assert(isa<GetElementPtrInst>(U) || isa<GEPOperator>(U));
+  auto FieldLoc = calculateFieldNumFromGEP(U);
+  if (FieldLoc == 0)
+    return;
+  for (auto *User : U->users()){
+    DEBUG_WITH_TYPE(DEBUG_TYPE_IR, dbgs() << "Check user of " << *U << ": " << *User << "\n");
+    if (isa<Instruction>(User)){
+      auto* Inst = dyn_cast<Instruction>(User);
+      if (Inst->getOpcode() == Instruction::Load)
+        addFieldAccessNum(Inst, FieldLoc);
+      else if (Inst->getOpcode() == Instruction::Store){
+        if (U == Inst->getOperand(1))
+          addFieldAccessNum(Inst, FieldLoc);
+      }
+      else{
+        if (Inst->getOpcode() == Instruction::Call || Inst->getOpcode() == Instruction::Invoke){
```

Comment:
I just removed this part in a commit before. I figured it could be a problem. I don't why it's not updated...